projects/igniteui-angular/src/lib/grids/common/pipes.ts

```diff
@@ -355,3 +355,39 @@ export class IgxColumnFormatterPipe implements PipeTransform {
         return formatter(value);
     }
 }
+
+@Pipe({
+    name: 'gridAddRow',
+    pure: true
+})
+export class IgxGridAddRowPipe implements PipeTransform {
+
+    constructor(private gridAPI: GridBaseAPIService<IgxGridBaseDirective & GridType>) { }
+
+    transform(collection: any, parentID: number, enableAddRow: boolean, cancelAddMode: boolean, pipeTrigger: number) {
+        const grid = this.gridAPI.grid;
+        if (!enableAddRow) {
+            return collection;
+        }
+        if (parentID === -1) {
+            const result = collection.map(record => {
+                return grid.isAddRowRecord(record) ? record.recordRef : record;
+            });
+            grid.data = result;
+            return result;
+        }
+        const parentRecordIndex = collection.findIndex(record => record[grid.primaryKey] === parentID);
+        if (cancelAddMode) {
+            collection.splice(parentRecordIndex + 1, 1);
+            return collection;
+        }
+        const row = {...collection[parentRecordIndex]};
+        Object.keys(row).forEach(key => row[key] = undefined);
+        const rec = {
+            recordRef: row,
+            addRow: true
+        };
+        collection.splice(parentRecordIndex + 1, 0, rec);
```

Comment:
Since this pipe runs after all other pipes, there may be cases when the row should not be shown below the parent record index. For example, if you have a grid with master-detail views, if the new row is displayed under the parent (master row) it would visually break the hierarchy since the new row will appear above the details view:
![image](https://user-images.githubusercontent.com/10397980/92750217-8a09ea80-f38f-11ea-98ed-caa4868008e2.png)

In those cases it may be better to show the add row above it.
Also the add row should not allow expand/collapse operations